google/cloud/aiplatform/jobs.py

```diff
@@ -1219,13 +1219,60 @@ def from_local_script(
             staging_bucket=staging_bucket,
         )
 
+    def web_access_uris(self):
+        if self.state in _JOB_COMPLETE_STATES:
+            self._log_job_state()
+            _LOGGER.info(
+                "Access to the interactive shell terminals are only available when the job is running."
+            )
+            return
+        self._web_access_uris()
+
+    def _web_access_uris(self):
+        """Helper method to check on job to get web access uris.
+        """
+
+        # Used these numbers so failures surface fast
+        wait = 5  # start at five seconds
+        log_wait = 5
+        max_wait = 60 * 1  # 1 minute wait
+        multiplier = 2  # scale wait by 2 every iteration
+
+        previous_time = time.time()
+        while (
+            self.state not in _JOB_RUNNING_STATE
+            and self.state not in _JOB_COMPLETE_STATES
+        ):
+            current_time = time.time()
+            if current_time - previous_time >= log_wait:
+                self._log_job_state()
+                log_wait = min(log_wait * multiplier, max_wait)
+                previous_time = current_time
+            time.sleep(wait)
+
+        if self.state in _JOB_RUNNING_STATE:
+            _LOGGER.info(
+                "%s %s access the interactive shell terminals for this job at the following links:\n%s"
+                % (
+                    self.__class__.__name__,
+                    self._gca_resource.name,
+                    "\n".join(
+                        [
+                            "%s:\n%s" % (worker, uri)
+                            for worker, uri in self._gca_resource.web_access_uris.items()
```

Comment:
Can we rely on the presence of these uris in the resource proto as the condition to log them so we don't have to maintain more than one polling block?

Something like this in the main polling block:
```Python
  has_logged_web_access_uri = False
  while self.state not in _JOB_COMPLETE_STATES:
      current_time = time.time()
      if current_time - previous_time >= log_wait:
          self._log_job_state()
          log_wait = min(log_wait * multiplier, max_wait)
          previous_time = current_time
      if self._gca_resource.web_access_uris and not has_logged_web_access_uris:
        self._log_web_access_uris()
      time.sleep(wait)
```

Reply:
Do you mean the `def _block_until_complete(self)` from `_Job` class as the main polling block? I was going with modifying the `def _block_until_complete(self)` utilizing the `web_access_uris` present in resource proto initially. Here are the reasons I decided not to:
1) It works well for CustomJob class when modifying the `def _block_until_complete(self)`. However, for `HyperparameterTuningJob`, the pulling is needed at the trial level instead of the job level, so we will need a customized `def _block_until_complete(self)` for HP class. 
2) For `Custom*Training` classes, their `def _block_until_complete(self)` is at training pipeline level, where a separate pulling to its `backingCustomJob` will be needed, and it will be outside the Job's `def _block_until_complete(self)`. 
So all these will end up having more duplicated logic around `def _block_until_complete(self)`. I figure it will be cleaner to have 2 pulling block but with different function, and they can be reused whenever is needed.